### PR TITLE
Refactor normal forms in strings

### PR DESCRIPTION
**Pull request by ajreynol**
This substantially refactors the core solver in TheoryStrings so that the notion of normal forms is encapsulated in a data structure.

It moves all core functions related to normal form representation/explanation to normal_form.h/cpp.

It documents all the core functions in TheoryString regarding the normal form inference schema apart from loop detection/processing.

The behavioral differences in this PR are intended to be minor.

### CODE REVIEW

**Comment by ajreynol** [comment]
This is further work towards #1881.

File: src/theory/strings/theory_strings.cpp [hunk 1]
```diff
@@ -2793,41 +2818,56 @@ void TheoryStrings::getNormalForms( Node &eqc, std::vector< std::vector< Node >
...
-    normal_forms_exp.push_back( std::vector< Node >() );
-    normal_forms_exp_depend.push_back( std::map< Node, std::map< bool, int > >() );
+    // This case happens when there are no non-trivial normal forms for this
+    // equivalence class. For example, given assertions:
+    //   { x = y ++ z, x = y, z = "" }
+    // The equivalence class of { x, y, y ++ z } is such that the normal form
+    // of all terms is a variable (either x or y) in the equivalence class
+    // itself. Thus, the normal form of this equivalence class can be assigned
+    // to one of these variables.
+    // We use a non-concatentation term among the terms in this equivalence
```
Comment (4tXJ7f) [inline src/theory/strings/theory_strings.cpp]:
```suggestion
    // We use a non-concatenation term among the terms in this equivalence
```

File: src/theory/strings/theory_strings.cpp [hunk 2]
```diff
@@ -2625,163 +2640,173 @@ void TheoryStrings::normalizeEquivalenceClass( Node eqc ) {
...
-      if( cmp==(k==1) ){
-        nf_exp_depend_n[exp][k==1] = val;
-      }
-    }
+NormalForm& TheoryStrings::getNormalForm(Node n)
+{
+  std::map<Node, NormalForm>::iterator itn = d_normal_form.find(n);
+  if (itn == d_normal_form.end())
+  {
+    Trace("strings-warn") << "WARNING: returning empty normal form for " << n
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/theory_strings.cpp]: I don't want this to be printed to the user, so I don't want Warning() here.

File: src/theory/strings/normal_form.h [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+   * in reverse order.
+   */
+  void reverse();
+  /** split constant
+   *
+   * Splits the constant in d_nf at index to constants c1 and c2. The flag
+   * isRev indicates whether the normal form has been reversed (this impacts
+   * how the dependency indices are updated).
+   */
+  void splitConstant(unsigned index, Node c1, Node c2, bool isRev);
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/strings/normal_form.h]: Done

**Pull request merged**